### PR TITLE
Add support for direct to parquet files

### PR DESCRIPTION
**Pull request by robotblake**
This is super gross and doesn't have tests (beyond my own local testing) but appears to work.  I ended up just scrapping a whole bunch of work and made the new code path only handle direct-to-parquet files so this could get pushed out.

### CODE REVIEW

**Comment by fbertsch** [comment]
Can you describe what you tested this on? Just so I can get an idea of what's been covered.

File: parquet2hive_modules/parquet2hivelib.py [hunk 1]
```diff
@@ -263,3 +299,96 @@ def _get_common_prefixes(bucket, prefix=''):
...
+    if partitions:
+        columns = ", ".join("`{}` string".format(p) for p in partitions)
+        partition_decl = " partitioned by ({})".format(columns)
+    else:
+        partition_decl = ""
+
+    return "drop table if exists {0}; create external table {0}({1}){2} stored as parquet location '\"'{3}'\"'; msck repair table {0};".format(table_name, fields_decl, partition_decl, location)
+
+
+def parquet2sql_fields(schema, children):
```
Comment (fbertsch) [inline parquet2hive_modules/parquet2hivelib.py]: The docs say that `schema` is a list of `SchemaElements`, which actually represents the tree but in depth-first traversal. Isn't this function assuming it's breadth first traversal? I.e. for
```
   p
  / \
 c1 c2
 /
c3
```
it assumes `[p, c1, c2, c3]`, but depth-first traversal would be  `[p, c1, c3, c2]`

File: parquet2hive_modules/parquet2hivelib.py [hunk 2]
```diff
@@ -263,3 +299,96 @@ def _get_common_prefixes(bucket, prefix=''):
...
+        if data_type is not None:
+            dictobj.append((elem.name, data_type))
+        else:
+            raise Exception("can't convert field: {}".format(elem))
+    return dictobj
+
+
+def parquet2sql_column(field_name, data_type, in_struct=False):
+    spacer = ":" if in_struct else ""
+    if data_type[0] == "array":
```
Comment (fbertsch) [inline parquet2hive_modules/parquet2hivelib.py]: I see this repeated in 3 functions. Please pull it out to a separate function

File: parquet2hive_modules/parquet2hivelib.py [hunk 3]
```diff
@@ -263,3 +299,96 @@ def _get_common_prefixes(bucket, prefix=''):
...
+        data_type = parquet2sql_map(data_type[1])
+    elif data_type[0] == "struct":
+        data_type = parquet2sql_struct(data_type[1])
+    else:
+        data_type = data_type[0]
+    return "array<{}>".format(data_type)
+
+
+def parquet2sql_map(data_type):
+    key_type = data_type[0][1][1][1][0][1][0]
```
Comment (fbertsch) [inline parquet2hive_modules/parquet2hivelib.py]: Can you document the reason there is this path to the key/value?

**Review by fbertsch — CHANGES_REQUESTED** [review]

**Comment by fbertsch** [comment]
@robotblake we should make it a priority to include tests for this. Trink would probably be happy to help make a direct-to-parquet output with most/all data types that we could test on.

File: tests/parquet2hive_test.py [hunk 4]
```diff
@@ -561,10 +562,358 @@ def test_fail_on_too_small(self):
...
+    def test_new_dataset_schema(self):
+        _setup_module()
+
+        obj = bucket.Object('my/new-dataset')
+        with open(new_dataset_file, 'rb') as fileobj:
+            obj.upload_fileobj(fileobj)
+
+        schema = lib.read_parquet_schema(obj)
+        assert lib.build_tree(schema[1:], schema[0].num_children) == NEW_DATASET_TREE
+
```
Comment (fbertsch) [inline tests/parquet2hive_test.py]: No tree for the complex schema? ;)

File: parquet2hive_modules/parquet2hivelib.py [hunk 5]
```diff
@@ -263,3 +317,60 @@ def _get_common_prefixes(bucket, prefix=''):
...
+        else:
+            return 'array<{}>'.format(sql_type(elem[4][0], False))
+    if elem[1] == 'group' and elem[3] in ('map', 'map_key_value'):
+        if elem[4][0][0] == 'repeated' and elem[4][0][1] == 'group':
+            if len(elem[4][0][4]) == 2:
+                return 'map<{},{}>'.format(sql_type(elem[4][0][4][0]), sql_type(elem[4][0][4][1]))
+    if elem[1] == 'group' and elem[3] is None or elem[3] == -1:
+        subs = ['`{}`: {}'.format(sub[2], sql_type(sub, False)) for sub in elem[4]]
+        return 'struct<{}>'.format(', '.join(subs))
+    if elem[0] == 'repeated' and root:
```
Comment (fbertsch) [inline parquet2hive_modules/parquet2hivelib.py]: A special case for root-level repeated elements? That seems... odd.

Comment (robotblake) [reply parquet2hive_modules/parquet2hivelib.py]: Yeah, that does look wrong, let me take a look.

**Comment by robotblake** [comment]
I added a pretty in-depth test case and a ton of tests, I also tested that the SQL generated by the old methods match the SQL generated by the new method against all of the datasets in the telemetry-parquet bucket.

File: parquet2hive_modules/parquet2hivelib.py [hunk 6]
```diff
@@ -204,62 +219,76 @@ def avro2sql(avro, name, version, location, partitions, with_version=True):
...
+    for _ in range(children):
+        elem = schema.pop(0)
+
+        elem_type = 'group' if elem.type is None else Type._VALUES_TO_NAMES[elem.type].lower()
+        converted = None if elem.converted_type is None else ConvertedType._VALUES_TO_NAMES[elem.converted_type].lower()
+        repetition = FieldRepetitionType._VALUES_TO_NAMES[elem.repetition_type].lower()
+
+        if elem_type == 'group':
+            subs = build_tree(schema, elem.num_children)
+            retval.append((repetition, elem_type, elem.name, converted, subs))
```
Comment (fbertsch) [inline parquet2hive_modules/parquet2hivelib.py]: Why not have this instead append a dict, and label each element (e.g. {"repetition": repetition, "type": elem_type, etc.}). Would make this, and the rest of the code, more readable.

File: parquet2hive_modules/parquet2hivelib.py [hunk 7]
```diff
@@ -204,62 +219,76 @@ def avro2sql(avro, name, version, location, partitions, with_version=True):
...
+        else:
+            retval.append((repetition, elem_type, elem.name, converted))
+
+    return retval
+
+
+def sql_type(elem):
+    if elem[1] == 'group' and elem[3] == 'list' and len(elem[4]) == 1:
```
Comment (fbertsch) [inline parquet2hive_modules/parquet2hivelib.py]: Can you put some comments here describing these structures and control flow?

File: tests/parquet2hive_test.py [hunk 8]
```diff
@@ -423,124 +427,365 @@ def test_ignore_dir(self):
...
+            + ") stored as parquet location '\"'s3://test-bucket/dataset.parquet'\"'; " \
+            + "msck repair table dataset_table;"
+
+NEW_DATASET_SQL = "drop table if exists new_dataset_table; " \
+                + "create external table new_dataset_table(" \
+                    + "`id` bigint" \
+                + ") stored as parquet location '\"'s3://test-bucket/new-dataset.parquet'\"'; " \
+                + "msck repair table new_dataset_table;"
+
+COMPLEX_SQL = "drop table if exists complex_table; " \
```
Comment (fbertsch) [inline tests/parquet2hive_test.py]: I think I would prefer these moved to a separate .sql file

**Review by fbertsch — CHANGES_REQUESTED** [review]

**Comment by fbertsch** [comment]
@robotblake have you had a chance to test this on the new direct-to-parquet versioned datasets?

**Review by fbertsch — APPROVED** [review]

**Pull request merged**